File: intrinsic/platform/pubsub/python/kvstore_test.py
# Copyright 2026 Intrinsic Innovation LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Tests for the KeyValueStore Python bindings."""

from absl.testing import absltest

from intrinsic.platform.pubsub.python import pubsub


class KVStoreTest(absltest.TestCase):

  def test_make_key(self):
    self.assertEqual(
        pubsub.KeyValueStore.MakeKey("/foo", "bar", "baz/"), "foo/bar/baz"
    )
    self.assertEqual(
        pubsub.KeyValueStore.MakeKey("foo", "bar", "baz"), "foo/bar/baz"
    )
    self.assertEqual(
        pubsub.KeyValueStore.MakeKey("///foo", "bar///", "///baz///"),
        "foo/bar/baz",
    )
    self.assertEqual(
        pubsub.KeyValueStore.MakeKey("/foo/", "/bar/", "/baz/"), "foo/bar/baz"
    )
    self.assertEqual(pubsub.KeyValueStore.MakeKey("foo", "", "bar"), "foo/bar")
    self.assertEqual(
        pubsub.KeyValueStore.MakeKey("foo", "///", "bar"), "foo/bar"
    )
    self.assertEqual(pubsub.KeyValueStore.MakeKey("///", "///", "///"), "")
    self.assertEqual(pubsub.KeyValueStore.MakeKey(), "")
    self.assertEqual(
        pubsub.KeyValueStore.MakeKey("foo/bar", "baz"), "foo/bar/baz"
    )

  def test_set_with_verification_options(self):
    mode = pubsub.SetWithVerificationOptions.VerificationMode
    self.assertCountEqual(
        [m.name for m in mode], ["FIRST_REPLY", "HIGH_CONSISTENCY"]
    )

    options = pubsub.SetWithVerificationOptions()
    self.assertEqual(options.mode, mode.HIGH_CONSISTENCY)
    self.assertEqual(options.timeout, 30.0)

    options.mode = mode.FIRST_REPLY
    options.timeout = 5.0
    self.assertEqual(options.mode, mode.FIRST_REPLY)
    self.assertEqual(options.timeout, 5.0)


if __name__ == "__main__":
  absltest.main()

File: intrinsic/platform/pubsub/golang/kvstore_test.go
// Copyright 2026 Intrinsic Innovation LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package kvstore

import (
	"testing"
)

func TestMakeKey(t *testing.T) {
	tests := []struct {
		name     string
		parts    []string
		expected string
	}{
		{
			name:     "some parts without slashes and some without",
			parts:    []string{"/foo", "bar", "baz/"},
			expected: "foo/bar/baz",
		},
		{
			name:     "no slashes present",
			parts:    []string{"foo", "bar", "baz"},
			expected: "foo/bar/baz",
		},
		{
			name:     "multiple slashes",
			parts:    []string{"///foo", "bar///", "///baz///"},
			expected: "foo/bar/baz",
		},
		{
			name:     "single slashes on both sides",
			parts:    []string{"/foo/", "/bar/", "/baz/"},
			expected: "foo/bar/baz",
		},
		{
			name:     "empty part in middle",
			parts:    []string{"foo", "", "bar"},
			expected: "foo/bar",
		},
		{
			name:     "slashes only part in middle",
			parts:    []string{"foo", "///", "bar"},
			expected: "foo/bar",
		},
		{
			name:     "only slashes across all parts",
			parts:    []string{"///", "///", "///"},
			expected: "",
		},
		{
			name:     "no parts",
			parts:    []string{},
			expected: "",
		},
		{
			name:     "internal slash preserved",
			parts:    []string{"foo/bar", "baz"},
			expected: "foo/bar/baz",
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			actual := MakeKey(tc.parts...)
			if actual != tc.expected {
				t.Errorf("MakeKey() = %v, expected %v", actual, tc.expected)
			}
		})
	}
}
